File: go/min_stack.go
package main

import "fmt"

type MinStack struct {
	vals []int
	mins []int
}

func Constructor() MinStack {
	return MinStack{
		vals: []int{},
		mins: []int{},
	}
}

func (this *MinStack) Push(val int) {
	var minVal int
	if len(this.mins) == 0 {
		this.mins = append(this.mins, val)
	} else {
		minVal = min(val, this.mins[len(this.mins)-1])
		this.mins = append(this.mins, minVal)
	}

	this.vals = append(this.vals, val)
}

func (this *MinStack) Pop() {
	this.vals = this.vals[:len(this.vals)-1]
	this.mins = this.mins[:len(this.mins)-1]
}

func (this *MinStack) Top() int {
	return this.vals[len(this.vals)-1]
}

func (this *MinStack) GetMin() int {
	return this.mins[len(this.mins)-1]
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */
func main() {
	obj := Constructor()
	obj.Push(-1)
	obj.Push(0)
	obj.Push(-2)
	obj.Pop()
	obj.Top()
	obj.GetMin()

	fmt.Println(obj.vals)
	fmt.Println(obj.mins)
}

File: python/min_stack.py
class MinStack:

    def __init__(self):
        self.items = []
        self.min = []

    def push(self, val):
        self.items.append(val)
        val = min(val, self.min[-1] if self.min else val)
        self.min.append(val)

    def pop(self):
        self.items.pop()
        self.min.pop()

    def top(self):
        return self.items[-1]

    def getMin(self):
        return self.min[-1]

tmp = MinStack()
tmp.push(1)
tmp.push(-2)
tmp.pop()
print(tmp.top())
print(tmp.getMin())